COORDINATION.md

```diff
@@ -0,0 +1,103 @@
+# Project Coordination Guidelines
+
+Arguably, communication is our primary concern throughout the entire process of
+designing, implementing, documenting, testing, and disseminating our programs.
+But, without question, collaboration on software development calls for a lot of
+communication.
+
+This document records our shared (always evolving) understanding of the key ways
+we communicate in furtherance of our collaboration. Its intention is to ensure that we are aligned between ourselves in regard to how and why we use the tools we do.
+
+## Elements of our project coordination
+
+We currently use GitHub as the main platform for sharing the communications
+that let us coordinate our work. GitHub has a number of different features for
+enabling cooperative work, but it's not necessarily obvious what the scope or
+intended meaning of each feature is. The following table is used to align our
+shared understanding of how to use and interpret the main features:
+
+| Feature                  | Scope                                                   | Communicates...        |
+|--------------------------|---------------------------------------------------------|------------------------|
+| [issue][issues]          | a distinct problem or task                              | what to work on        |
+| [pull request][prs]      | (part of) a solution or fulfillment of a task           | the ongoing work       |
+| [milestone][milestones]  | a distinct feature or a closely related set of problems | what work is towards   |
+| [label][labels]          | a more or less vague category/topic/quality             | attributes of the work |
+| [project board][project] | concurrent human processes with a common dynamic        | how work is done       |
+| [discussion][discussion] | free ranging and open discussion                        | whatever               |
+
+(For more on projects vs. milestones, see
+https://stackoverflow.com/a/47542346/1187277)
+
+[issues]: https://github.com/informalsystems/apalache/issues
+[prs]: https://github.com/informalsystems/apalache/pulls
+[milestones]: https://github.com/informalsystems/apalache/milestones
+[labels]: https://github.com/informalsystems/apalache/issues/labels
+[project]: https://github.com/orgs/informalsystems/projects/30
+[discussion]: https://github.com/informalsystems/apalache/discussions
+
+### Project board
+
+Our project board is the main resource to help with our day-to-day coordination. We use a Kanban-styled continuous workflow board with a high level view of whats being done and some queues for what is up next.
```

Comment:
We should still discuss if this type of continuous workflow board suits the entire team. Jure expressed that he likes to see some reduction in the amount of work to do (like a burndown chart), but if we are continuously moving cards to done and continuously adding tasks, we lose any relation between our capacity and the amount of work to do. This type of visualization requires that we compromise with a certain amount of work inside a time-box where we won't change what we are compromised to do (i.e. add new tasks), like in a sprint in scrum-based approches.

Personally, I believe that the main problems in compromising with an amount of work are
1. We lose flexibility
2. We have to work with some sort of time-box (otherwise we can just change what we are compromised to at any given time and it becomes continuous again). Time-boxes can be unfounded sources of pressure and disappointment if they are not treated 100% maturely. It's usually very hard for people to distinguish a time-box like this, that has a specific purpose, from a imposed deadline.

Reply:
I think we can get signal on some of this via existing features. For example, we can use milestones with due dates to get time-limited progression on a certain set of issues. If we want a view of our throughput, we can filter all issues/prs closed within a certain date range (I think!).

iiuc your point here, I agree that we shouldn't expect that the kanban boards to give us this kind of perspective, and that the costs trying to work within time boxes are pretty high.

Reply:
I agree with what was said, but would like to point two things out.

1. iirc the point of having different buckets on the project board was to see how the team's time was spent in different types of tasks, i.e., we spent a lot of effort recently on refactoring, so let's put more focus on usability next week. This information is lost if we keep refiling the buckets as Gabriela said. I always saw the buckets as a selection of issues without time constraints i.e., once we get all this done we discuss what to do next, but maybe it was just me.
1. I don't think we should use milestones with two different meanings, or at least not without a very clear agreement on how to do so.

Reply:
About 1., maybe we can just use labels and look at the recent labels on the "Done" column? If we are reflecting on the past (i.e. last week), the "Done" column should be our best source of information. We could event have buckets for the Done side such as "Done Usability", "Done Refactoring" if we think that's better than using tags.

Reply:
Here's the way I think the project board should be used:
To start, we have a [plan][https://github.com/informalsystems/strategy/blob/igor/yearly2021-2022/projects/apalache/yearly2021-2022/plan2022.md] for 2022.
The plan is not absolute, and obviously can't predict spontaneous issues or ideas that come up, but it does act as a partial order on categories of tasks.
In that context, at least to me, the project board acts as the refinement of tasks for a fixed period of time (e.g. Q4 or one month of Q4).
Therefore, until the end of the fixed time interval, the number of tasks should monotonically decrease. This does not mean that the fixed time window is a deadline, since a task can be kept on the board after the window (deciding that is what meetings are for).
The purpose of monthly or quarterly meetings is then to decide what tasks to put on, or remove from, the board.
The purpose of weekly meetings is to evaluate whether some buckets on the boards are being emptied much faster than others, with the ultimate goal being that "less interesting" (according to subjective perspective) tasks (e.g. documentation, testing, MBT) don't get neglected.

I don't think milestones/labels are particularly relevant in this scenario (arguably, milestones are subsets of labels and buckets contain subsets of milestones).

Adressing concerns:
>  We lose flexibility

Yes, in the sense that you lose the ability to arbitrarily pick your next issue from outside the board. Not really, in the sense that the board is not immutable, and if there is a sudden need to start working on something not planned, there is a meeting at least once every week, during which we can add things to the board. I'll also mention that critical bugfixes are understood to be outside the board and can (should?) be fixed asap.

> We have to work with some sort of time-box (otherwise we can just change what we are compromised to at any given time and it becomes continuous again). Time-boxes can be unfounded sources of pressure and disappointment if they are not treated 100% maturely. It's usually very hard for people to distinguish a time-box like this, that has a specific purpose, from a imposed deadline.

Yes, that's what quarterly planning has always been. Time-boxes are only pressure if there exists some entity imposing consequences on non-adherence. Currently, the only source of external feature constraints (that aren't just requests by other teams at Informal) is the funding agency that is paying for the new Apalache project (which is to say, this is mostly a non-issue IMO, we don't have traditional deadlines for concrete deliverables).

Reply:
@Kukovec In this proposal, this feature is handled by milestones:

- > until the end of the fixed time interval, the number of tasks should monotonically decrease

Note that milestones are very well equipped for meeting this need, as you can specify a deadline, track the proportion of finished vs. unfinished tickets, and each milestone has the same logic to support this kind of thing. 

> arguably, milestones are subsets of labels and buckets contain subsets of milestones

You could of course collapse all the distinction between the different feature sets and view everything as a "mode of label", but what is gained? Milestones have distinct features not supported by labels (deadlines, progress meters), and their intended use is quite orthogonal to general labels. The columns in projects boards area really intended to track *status* of issues/prs. A single work items progresses through these statuses. It gives us a temporal view of the process by which work is done. The current board *does* use the status columns to group things in a way that is redundant with the new milestones, but don't think it adds anything, except making the board more chaotic, because the same x axis is now encoding two different kinds of information simultaneously: temporal status of work and the "bucket" within which the work falls.

Perhaps we could have a sync discussion where we compare the old and new boards, and look at what milestones actually offer, so we can see how to meet your needs here?

It is possible too that the kind of metrics you want are best served by doing some report generation on tickets within a given timeframe. That should be quite easy to rig up, whether by pulling in a SasS integration or just by posing some queries to github.

Reply:
@rodrigo7491 

> iirc the point of having different buckets on the project board was to see how the team's time was spent in different types of tasks, i.e., we spent a lot of effort recently on refactoring, so let's put more focus on usability next week. This information is lost if we keep refiling the buckets as Gabriela said. I always saw the buckets as a selection of issues without time constraints i.e., once we get all this done we discuss what to do next, but maybe it was just me.

These buckets have moved into milestones and/or labels. This information is not lost, it's just not represented in the **status** columns (because those columns are designed to reflect status). All this info on different buckets is still present, and readily available through filters and views.

> I don't think we should use milestones with two different meanings, or at least not without a very clear agreement on how to do so.

Where do you see milestones being used with two different meanings?

Reply:
@shonfeder 

> These buckets have moved into milestones and/or labels. This information is not lost, it's just not represented in the **status** columns (because those columns are designed to reflect status). All this info on different buckets is still present, and readily available through filters and views.

That makes sense, thanks!

> Where do you see milestones being used with two different meanings?

Not being used, but being suggested to be used with time constraints in https://github.com/informalsystems/apalache/pull/1184#discussion_r772595598. Maybe I misinterpreted what you meant there.

Reply:
I see. So my view is that this is a single way of using milestones: to track many different items of work that all build towards a (relatively) clearly delimited goal, and in particular I think this goal should have a well-defined end state. 

But milestones support deadlines. So we can either add a deadline, indicating that we are tracking work towards a goal which we aim to complete within a specific timeframe, or we can leave it without a deadline, in which case completion of the goal isn't time constrained.

Reply:
Maybe to address the ambiguity around milestones we should add a section just to detail the features and proposed use of milestones... will think on this tommorrow.

Reply:
> Therefore, until the end of the fixed time interval, the number of tasks should monotonically decrease.

This is an idealistic point of view that assumes that all tasks can be planned in advance. They cannot. In reality, we often start with an issue and divide it into smaller issues as we identify them. Of course, it does not mean that we should not plan. We should. But instead of decomposing a single task into subtasks in advance, we can guess the complexity of the task, which would intuitively translate into the number of subtasks (issues) for the initial task. It's hard to guess the actual subtasks, but it's relatively easy to weigh a task.

That said, I find the number of tasks to be a bad metric. If we attached complexity to each task, we could say that the complexity of the milestone (as a sum of the task complexities) should decrease.